### ISSUE TITLE
Using with Debian rather than Ubuntu

### ISSUE THREAD

**Issue by sammcj** [progrium/dokku#976]
Is it possible to use Dokku with Debian rather than Ubuntu?

* Debian packages are generally more thoroughly tested
* Debian is lighter-weight than Ubuntu
* Debian has more mature, sensible defaults
* Debian (backports) generally has newer packages than those available in Ubuntu
* Debian Jessie (8) isn't far off being released and uses systemd

**Comment by josegonzalez** [comment]
We don't test Debian, but in theory it would work. We have a trusty repo available at packagecloud.com/dokku/dokku, can you let me know of it works on the latest stable Debian?

**Comment by josegonzalez** [comment]
@sammcj bump

**Comment by sammcj** [comment]
Hey, sorry crazy week at work - yup I'll give it a go!

**Comment by sammcj** [comment]
Ok first problem, the dokku deb is looking for 'lxc-docker' which I'm assuming is what Ubuntu calls Docker - which is interesting since Docker hasn't used lxc for quite some time.

```
root@ext-samm:~# apt-get install lxc-docker
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package lxc-docker is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'lxc-docker' has no installation candidate
```

The official Docker packages are called `docker.io`

```
root@ext-samm:~# dpkg -s docker.io
Package: docker.io
Status: install ok installed
Priority: optional
Section: admin
Installed-Size: 19199
Maintainer: Paul Tagliamonte <paultag@debian.org>
Architecture: amd64
Version: 1.3.3~dfsg1-2
Replaces: docker (<< 1.5~)
Depends: adduser, iptables, init-system-helpers (>= 1.18~), perl, libapparmor1 (>= 2.6~devel), libc6 (>= 2.14), libdevmapper1.02.1 (>= 2:1.02.90), libsqlite3-0 (>= 3.5.9)
Recommends: aufs-tools, ca-certificates, cgroupfs-mount | cgroup-lite, git, xz-utils
Suggests: btrfs-tools, debootstrap, lxc, rinse
Breaks: docker (<< 1.5~)
Conffiles:
 /etc/bash_completion.d/docker 7fcdf73ff720606870336942b7efcd74
 /etc/default/docker ce88dab1dcba6f92903120b9beba2521
 /etc/init.d/docker 870b6e52007eca22390f36e3b2547954
 /etc/init/docker.conf 4a2cb4601b893c7c3db45e4886960bdb
Description: Linux container runtime
 Docker complements kernel namespacing with a high-level API which operates at
 the process level. It runs unix processes with strong guarantees of isolation
 and repeatability across servers.
 .
 Docker is a great building block for automating distributed systems:
 large-scale web deployments, database clusters, continuous deployment systems,
 private PaaS, service-oriented architectures, etc.
 .
 This package contains the daemon and client. Using docker.io on non-amd64 hosts
 is not supported at this time. Please be careful when using it on anything
 besides amd64.
 .
 Also, note that kernel version 3.8 or above is required for proper operation of
 the daemon process, and that any lower versions may have subtle and/or glaring
 issues.
Built-Using: golang-context (= 0.0~git20140604.1.14f550f-1), golang-dbus (= 1-1), golang-go-patricia (= 1.0.1-1), golang-go-systemd (= 2-1), golang-go.net-dev (= 0.0~hg20131201-1), golang-gocapability-dev (= 0.0~git20140516-1), golang-gosqlite-dev (= 0.0~hg20130601-1), golang-mux (= 0.0~git20140505.1.136d54f-2), golang-pty (= 0.0~git20140315.1.67e2db2-1)
Homepage: https://github.com/docker/docker
```

**Comment by josegonzalez** [comment]
Mmm yeah, https://docs.docker.com/installation/ubuntulinux/ references lxc-docker. You *could* build your own package using `vagrant up build` after customizing the [control](https://github.com/progrium/dokku/blob/master/debian/control) file to reference dokku.io.

The docs on the docker page say that the official repo (docker-maintained) has an `lxc-docker` package, so try adding their apt repo before installing?

**Comment by sammcj** [comment]
OK, so turns out I *can* run `curl -sSL https://get.docker.com/ | sh` to install a package from Docker called lxc-docker (go figure).

Installed dokku from the 'trusty' repo and that seemed to go smoothly...
Just waiting on `Importing buildstep into docker (around 5 minutes)` now ... fingers crossed!

Do you use any upstart scripts for Dokku?
As Upstart is something that only Ubuntu uses, I'm testing this with Debian 8 (Jessie) which has systemd and is almost released as the current stable Debian version.

**Comment by josegonzalez** [comment]
Dokku doesn't need upstart, just docker. Dokku is basically a bunch of shell scripts that automates other commands. The only two things that need to run on your box are docker and nginx. I believe you'll want to run `dokku ps:restartall` when your server restarts to ensure all apps are being served properly, but thats about it.

**Comment by sammcj** [comment]
Ok seems to be up and working, I have noticed one bug - $HOME/VERSION is missing:

```
root@ext-samm:~# dokku version
cat: /home/dokku/VERSION: No such file or directory
Unable to determine dokku's version
```

```
root@ext-samm:/home/dokku# ls
HOSTNAME
```

**Comment by josegonzalez** [comment]
Yeah, bug in the package build script I mean to fix for the next release. Anything else funky?

**Comment by sammcj** [comment]
Yeah looks pretty lacking now that I'm poking around.
It like the dokku-installer script wasn't run after the package was installed (maybe related to this bug you mentioned?)

Running it manually fails to find the service:

```
root@ext-samm:~# cd /root/dokku && make dokku-installer
apt-get install -qq -y ruby
test -f /var/lib/dokku/.dokku-installer-created || gem install rack -v 1.5.2 --no-rdoc --no-ri
Fetching: rack-1.5.2.gem (100%)
Successfully installed rack-1.5.2
1 gem installed
test -f /var/lib/dokku/.dokku-installer-created || gem install rack-protection -v 1.5.3 --no-rdoc --no-ri
Fetching: rack-protection-1.5.3.gem (100%)
Successfully installed rack-protection-1.5.3
1 gem installed
test -f /var/lib/dokku/.dokku-installer-created || gem install sinatra -v 1.4.5 --no-rdoc --no-ri
Fetching: tilt-1.4.1.gem (100%)
Successfully installed tilt-1.4.1
Fetching: sinatra-1.4.5.gem (100%)
Successfully installed sinatra-1.4.5
2 gems installed
test -f /var/lib/dokku/.dokku-installer-created || gem install tilt -v 1.4.1 --no-rdoc --no-ri
Successfully installed tilt-1.4.1
1 gem installed
test -f /var/lib/dokku/.dokku-installer-created || ruby /root/dokku/contrib/dokku-installer.rb onboot
Installed Upstart service and default Nginx virtualhost for installer to run on boot.
test -f /var/lib/dokku/.dokku-installer-created || service dokku-installer start
Failed to start dokku-installer.service: Unit dokku-installer.service failed to load: No such file or directory.
Makefile:116: recipe for target 'dokku-installer' failed
make: *** [dokku-installer] Error 6
```
It looks like the service script was never installed:

```
root@ext-samm:~/dokku# locate dokku.service
root@ext-samm:~/dokku#
```

**Comment by josegonzalez** [comment]
Do you have a full log of the install process for everything?

**Comment by sammcj** [comment]
Ah I just spun up a temporary vm then spun it down again - I'm not overly stressed about getting Dokku going on Debian - but it would then be something I'd recommend people looking for Huroku alternatives.

**Comment by josegonzalez** [comment]
Heh, the service file isn't installed because there isn't one for systemd, just upstart. Here is the upstart version:

```
start on runlevel [2345]
exec /usr/local/share/dokku/contrib/dokku-installer.rb selfdestruct
```

Two questions:

- How do we detect usage of systemd vs upstart?
- What does the above script look like in systemd land?

I can probably modify the debian package to support both quite easily. It definitely doesn't work on Debian 7.x (nginx requirement is too low) so I'll try [this vm on vagrant](https://atlas.hashicorp.com/oar-team/boxes/debian8).

Should I try 7.8 with backports?

**Comment by sammcj** [comment]
I'd just test for weather it's Ubuntu or Debian (or just about anything else) - Ubuntu is actually ditching upstart in favor of SystemD soon too.

Debian 7 - Definitely use backports - it's a must have and very stable, 7.x currently has Nginx 1.6.2 in backports, anyone running Docker will most likely be using backports anyway as that gives Kernel 3.16 and much nicer BTRFS drivers.

Since 8 is just about to be released, I'd just get it working with that rather than worrying about 7 now, as a bonus that has Nginx 1.6 and will have 1.7 at some point.

**Comment by josegonzalez** [comment]
When 8 is released, then I can test against that, but as is I'm not going to try testing against something that isn't completely stable yet.

**Comment by ninive** [comment]
Hi guys, sorry I drop by, I can confirm Dokku 0.3.15 is working on Wheezy via this bash: 
https://gist.github.com/shirkey/9139468#file-install_dokku_to_debian_wheezy

I've just changed the last line to install the new version. Seems okay even installing plugins, as a side note mongoDb plugin has some problem with the mongoDB shell version, and i'm using it normally for testing everyday. As @sammcj pointed out, i love to have it on Debian.

Hope this helps! 
Cheers, 
Alex

**Comment by josegonzalez** [comment]
I'm going to say that we don't support debian installation until debian 8 comes out stable. Until then, we won't promote installation on it. There isn't any functional difference in your interaction with the system - you're deploying into a docker image - so Ubuntu should be fine for now, barring personal preference.

**Comment by ninive** [comment]
Hi José, 110% with you, make no sense to test and support before is considered stable.
Was for sure a good emphasis about Dokku working great and a little trick to try out a VM on the fly, nothing more.
As a plus, I was pushing some new stuff up in DigitalOcean today and the same random bug appear even in the Ubuntu Dokku droplet, same identical story, so it's not Debian.
I'm in touch with mongoDB author to try to understand and fix cause i would lile to push a new test app and i can't now ;)

Thanks a lot for your work on this, really appreciated, since the beginning.

**Comment by josegonzalez** [comment]
Yep, once Debian 8 is stable, I'll probably revisit, but for now we'll stick to Ubuntu as our sole officially supported OS.

**Comment by eljojo** [comment]
Hi,
I just tried installing on jessie and it failed. [Here](https://gist.github.com/eljojo/883518d6608507c10758)'s the output log.

The nginx repository is failing to install due to [this line](https://github.com/progrium/dokku/blob/02277cc411a368a19fefb43edbbaeb0804718785/plugins/nginx-vhosts/dependencies#L9).

I guess one question to answer is: what repositories to use in debian?
Surely docker and nginx right now have a decent version, but knowing debian's love for stability, we might want to refer to external repositories to get up-to-date versions of docker, nginx and others.

After simply removing the line, I get the [following error](https://gist.github.com/eljojo/be8a0feff34b4d02509b).

After removing the motd entry from [the dependencies file](https://github.com/progrium/dokku/blob/02277cc411a368a19fefb43edbbaeb0804718785/plugins/00_dokku-standard/install#L46-L58) the system *seems* to install properly:

```
dokku plugins-install
[ ok ] Starting nginx (via systemctl): nginx.service.
git describe --tags > ~dokku/VERSION  2> /dev/null || echo '~master (2015-05-04T17:41+0000)' > ~dokku/VERSION
```

What's also worth noting is that I found lots of lines [like this](https://github.com/progrium/dokku/blob/02277cc411a368a19fefb43edbbaeb0804718785/plugins/nginx-vhosts/dependencies#L4) using an $DOKKU_DISTRO variable.

There's probably something wrong somewhere related to git hooks, because I couldn't manage to push an app and deploy correctly.
Also the ``sshcommand acl-add dokku`` might not work in debian.

**Comment by eljojo** [comment]
After manually adding the git receive hook with:

```
root@dokku:/home/dokku# dokku apps:create rubytest
Creating rubytest... done
root@dokku:/home/dokku# su dokku
dokku@dokku:~$ ls
HOSTNAME  rubytest  VERSION  VHOST
dokku@dokku:~$ cd rubytest/
dokku@dokku:~/rubytest$ git init --bare .
Initialized empty Git repository in /home/dokku/rubytest/
dokku@dokku:~/rubytest$ cd hooks/
dokku@dokku:~/rubytest/hooks$ vim pre-receive
dokku@dokku:~/rubytest/hooks$ chmod +x pre-receive
```

I managed to push and deploy correctly!
There's probably lots of stuff still broken, but it's a good start :)

dokku version also works:

```
dokku@dokku:~$ dokku version
v0.3.17-65-g02277cc
```

**Comment by maethor** [comment]
In fact, you don't have to use the full bootstrap script. 0.3.13 deb package installs and works very well on Debian Jessie.

**Comment by josegonzalez** [comment]
Does anyone want to see if latest master will build packages that are capable of being installed on Debian Jessie?

**Comment by callahad** [comment]
TL:DR; the build / postinstall scripts make a lot of assumptions about their environment that don't necessarily hold for Debian. We're not ready to run there, yet.

Attempted to install v0.4.1 via packages from bootstrap.sh on Debian 8.1. Failures:

1. The herokuish postinstall requires sudo, but doesn't declare it in its dependencies. Sudo isn't necessarily available on Debian. This causes installation to fail. Workaround: apt-get install sudo;
2. The dokku postinstall runs core-plugins/enabled/00_dokku-standard/install which expects /etc/update-motd.d/ to exist, but it doesn't on Debian. This causes installation to fail. Workaround: mkdir /etc/update-motd.d/
3. The dokku postinstall attemps to chown something to `syslog:dokku`, but the `syslog` user does not necessarily exist on Debian.
4. The core plugin installation fails when it tries to apt-get update, since we install a bogus nginx source.

Attempt to make deb-all. Faillures:

1. The deb-setup target attempts to install ruby1.9.1-dev, which isn't available. ruby-dev appears to work.
2. The deb-herokuish target fails to build since it can't clone herokuish over ssh. Switching to https works.
3. The deb-dokku target fails if /vagrant isn't a directory
4. The deb-dokku target fails to locate the dokku.deb and move it unless $PWD is /vagrant
5. The deb-plugn target fails since it can't clone over ssh. Switching to https works.
6. Ditto for deb-sshcommand

**Comment by josegonzalez** [comment]
- Can we drop the sudo in the herokuish postinstall? What is it used for?
- Makes sense, what do other packages that want to provide an MOTD do?
- Heh, wtf do you guys use for the logging user?
- It's def not bogus! What version of nginx comes on Debian? We can probably skip that dep on non-ubuntu instances.

**Comment by callahad** [comment]
It's bogus in the sense that it points to an Ubuntu PPA but looks for a "jessie" dist, which, naturally, doesn't exist. :)

Debian is currently shipping 1.6.2 (jessie) and 1.9.4 (jessie-backports). Ubuntu is currently shipping 1.4.6 (trusty), and the unofficial PPA we use ships 1.9.4.

Nginx [ships](http://nginx.org/en/linux_packages.html) Debian and Ubuntu packages for both 1.8.0 (stable) and 1.9.5 (mainline).

...so you have your pick of those. What's the minimum version Dokku needs? ;)

**Comment by josegonzalez** [comment]
1.9.5. What gives us that?

**Comment by michaelshobbs** [comment]
I think you mean stable `1.8.0`

**Comment by josegonzalez** [comment]
Fine, 1.8.0 :(

**Comment by josegonzalez** [comment]
Well, this all seems to work well enough. I didn't change the nginx version on debian because debian users are more antsy about stuff like that. Going to merge as soon as tests pass and the next version of dokku should support debian with the same packaging :)

**Comment by josegonzalez** [comment]
The next released version will have experimental support for debian systems.
